TFM: Rework TFM Minimal configuration

joerchan opened the pull request:
Rework the TF-M minimal configuration. All features except from RNG are
disabled in this configuration.

Disables Internal Trusted Storage.
Disables Hash Crypto service.
Disables all Mbed Crypto options.
Minimal size reduced to 32KB.

Allow the CC3XX entropy driver to be disabled with TF-M when the
Crypto partition is disabled.
This would otherwise cause a compilation error.

Make the TFM hello world sample compile under different configurations.
The Crypto modules RNG and HASH can be disabled, in addition to the
entire crypto and platform partitions.

~Draft status:~
~This PR (and dependent PR) is on top of https://github.com/nrfconnect/sdk-nrf/pull/6698~

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-6755
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

tejlmand commented on modules/trusted-firmware-m/Kconfig.mbedtls_minimal.defconfig:
```diff
@@ -47,3 +47,192 @@ config MBEDTLS_POLY1305_C
 
 config MBEDTLS_PSA_CRYPTO_STORAGE_C
 	default n
+
```
why double spacing ?

tejlmand commented on modules/trusted-firmware-m/Kconfig.mbedtls_minimal.defconfig:
```diff
@@ -47,3 +47,192 @@ config MBEDTLS_POLY1305_C
+
+config MBEDTLS_LEGACY_CRYPTO_C
+	default n
+
```
why double spacing ?

tejlmand commented on modules/trusted-firmware-m/Kconfig.mbedtls_minimal.defconfig:
```diff
@@ -47,3 +47,192 @@ config MBEDTLS_POLY1305_C
+config PSA_WANT_ALG_SHA_256
+	bool
+	default n
+
```
why triple spacing ?

tejlmand commented on modules/trusted-firmware-m/Kconfig.mbedtls_minimal.defconfig:
```diff
@@ -47,3 +47,192 @@ config MBEDTLS_POLY1305_C
+config PSA_WANT_KEY_TYPE_DERIVE
+	bool
+	default n
+config PSA_WANT_KEY_TYPE_HMAC
```
maybe move some off the saved newlines down here :wink:

joerchan replied:
I have made an equal share redistribution of the newlines.

tejlmand commented on drivers/entropy/Kconfig:
```diff
@@ -5,7 +5,7 @@
 
 config ENTROPY_CC3XX
-	bool "Arm CC3XX RNG driver for Nordic devices" if !BUILD_WITH_TFM
+	bool "Arm CC3XX RNG driver for Nordic devices"
```
Reading the commit message and looking at this code.
> Allow the CC3XX entropy driver to be disabled with TF-M when the
> Crypto partition is disabled.
> This would otherwise cause a compilation error.

then i'm not sure I fully understand.

Now the cc3xx entropy driver can always be enabled, also when `TFM_PARTITION_CRYPTO=y`.
Based on the description in the commit message and this code change, I assume the following is now possible but will most likely result in a compilation error:
```
cmake ... -DCONFIG_BUILD_WITH_TFM=y -DCONFIG_TFM_PARTITION_CRYPTO=y -DCONFIG_ENTROPY_CC3XX=y
```

I don't believe this is intended.
Could you please check this, and update code or commit message if needed.

tejlmand replied:
right. seems I overlooked that code change, should be fine then.

Note, generally speaking, having a select directly to a specific driver can often be problematic, as it prevents alternative driver implementations, though in this particular case it is acceptable.

tejlmand replied:
@joerchan still I do not understand this part of the commit message ?
> Allow the CC3XX entropy driver to be disabled with TF-M when the Crypto partition is disabled.
> This would otherwise cause a compilation error.

Is it allowed to:
- Enable TF-M
- Disable tfm partition crypto
- Enable entropy cc3xx

Or what / when is the compilation error you refer to seen ?
I would like to understand the issue you are trying to fix, and thus be able to reproduce this issue and then see this PR actually solves this problem.

joerchan replied:
That would lead to a compilation error. The same one that I wanted to get rid of. 
```
entropy_cc310.c.obj): in function `entropy_cc3xx_rng_get_entropy':
entropy_cc310.c:42: undefined reference to `psa_generate_random'
```

The problem was really that ENTROPY_CC3XX was always enabled when TF-M is enabled. But the psa_generate_random is only available when TFM crypto partition with RNG support is enabled.

SPM also have the same problem but with SPM_SERVICE_RNG as a dependent configuration.

Having the select from SPM / TFM seems wrong to me now. Since SPM / TFM does not depend on this.
Instead the ENTROPY driver can use these services when they are enabled.

tejlmand replied:
> @tejlmand Trying to fix the above leads me down the rabbit hole of additional changes.

I can imagine.

> The commit itself is not needed to fix the TFM minimal configuration, so I have taken it out of this PR and will do a follow up PR later once I've figured out the rest.

Also note, that I will gladly approve even a minor improvement of minimizing the risk of getting compilation errors.
Just trying to ensure that if a commit is stating as fixing a particular issue, then it actually does fix it.
Otherwise it should just stated that it improves user-experience, cause then it's easier to decide which use-case is most important.

joerchan replied:
That's ok. The case I was trying to fix is really hard to get to, so not something that we have to address at the moment.

tejlmand reviewed (dismissed):
Generally looks good.

Some minor observations / questions.

SebastianBoe commented on Kconfig.nrf:
```diff
@@ -61,7 +61,7 @@ config ENTROPY_GENERATOR
 
 config PSA_WANT_ALG_CTR_DRBG
-	default y if ENTROPY_GENERATOR
+	default y if ENTROPY_GENERATOR && !TFM_MINIMAL
```
This is bad practice.

Features (DRBG) should not know of their users (TFM_MINIMAL).

Users should be enabling the features they need (which is evidently not DRBG in this case).

Can we add the removal of this patch to the DoD of https://projecttools.nordicsemi.no/jira/browse/NCSDK-13595 @frkv  ?

frkv replied:
This is default y for all things besides TF-M minimal configuration. This is the defacto PRNG both in CSPRNG in Zephyr and for The CryptoCell driver as well as the mbed TLS project, so no.

joerchan replied:
The reason for this being needed is that `Kconfig.nrf` is sourced after `Kconfig.mbedtls_minimal.defconfig`.

tejlmand replied:
@joerchan quite the opposite:
```h
menu "nrf (/projects/github/ncs/nrf)"
osource "/projects/github/ncs/nrf/Kconfig.nrf"
config ZEPHYR_NRF_MODULE
        bool
        default y
endmenu
menu "mcuboot (/projects/github/ncs/bootloader/mcuboot)"
osource "$(ZEPHYR_MCUBOOT_KCONFIG)"
config ZEPHYR_MCUBOOT_MODULE
        bool
        default y
endmenu
menu "mbedtls (/projects/github/ncs/mbedtls)"
osource "$(ZEPHYR_MBEDTLS_KCONFIG)"
config ZEPHYR_MBEDTLS_MODULE
        bool
        default y
endmenu
menu "trusted-firmware-m (/projects/github/ncs/modules/tee/tf-m/trusted-firmware-m)"
osource "$(ZEPHYR_TRUSTED_FIRMWARE_M_KCONFIG)"
config ZEPHYR_TRUSTED_FIRMWARE_M_MODULE
        bool
        default y
endmenu
....
```
https://github.com/nrfconnect/sdk-nrf/blob/2a2fe9c009f5dfda0e38c95afb67e92833f536cf/modules/modules.cmake#L8
https://github.com/nrfconnect/sdk-nrf/blob/2a2fe9c009f5dfda0e38c95afb67e92833f536cf/modules/trusted-firmware-m/Kconfig#L55

joerchan replied:
My mistake, then I don't have an explanation for why it is needed. The mbedtls_minimal_defconfig default is not set.

Maybe it is because `modules/mbedtls/Kconfig.psa` is sourced later, but that sets no default y.
```
At modules/mbedtls/Kconfig.psa:8
Included via /home/joakim/repos/ncs/zephyr/Kconfig:8 -> Kconfig.zephyr:33 -> modules/Kconfig:80 -> modules/mbedtls/Kconfig:176
```

tejlmand replied:
> My mistake, then I don't have an explanation for why it is needed.

Probably because the default that is sourced **first** wins.
This is also why a board default can overrule any other default.

joerchan replied:
I'll leave it as is, since I don't know how to fix this in a good practice way.

tejlmand replied:
I wonder, what's the reason for having this default in `Kconfig.nrf` in the first place, and not in nrfxlib, for example one of those two files:
https://github.com/nrfconnect/sdk-nrfxlib/blob/main/nrf_security/Kconfig
https://github.com/nrfconnect/sdk-nrfxlib/blob/main/nrf_security/Kconfig.psa

joerchan replied:
That is a good question. I'm also wondering why this default is there to begin with, in addition to the one above about ENTROPY_GENERATOR.
I'm  suspecting these two can actually be removed, since nrf_security has a select on ENTROPY_GENERATOR, and the PSA_WANT_ALG_CTR_DBRG is being select'ed from other places.
Possibly ENTROPY_CC3XX should have a select PSA_WANT_ALG_CTR_DRBG instead?

@frkv or @Vge0rge Any input on this?

Vge0rge replied:
I will have to test if this will work actually. I think one issue that we are trying to solve here is to set the correct defaults. The PSA_WANT configuration is placed in the mbedtls repo from zephyr. The entropy generator one is sourced from zephyr. So either we need downstream patches if we need to solve this or we need to make sure that the nrfxlib Kconfig is sourced first, which I would assume that is not possible.

frkv replied:
Combined knowledge about TF-M and Mbed TLS is somewhat spread around in multiple domains at the moment, so the configurations can be placed in multiple locations. What we do with the override is **very explicit**, so I argue that it isn't an anti-pattern. It is using default-overrides as they are expected to be used.

PSA_WANT_ALG_CTR_DRBG and PSA_WANT_ALG_HMAC_DRBG is an invention made by us, and it covers a feature that is not finalized inside the PSA Crypto specifications. There exists no driver-support for CTR_DRBG/HMAC_DRBG as well as no configuration to state with clear intent that you wish to provide (and use) one or the other. We want to upstream these config-additions (once PSA driver support is in place).

CTR_DRBG has a much higher impact (in code) than HMAC_DRBG. But more importantly CTR_DRBG is also the default solution for PRNG in our CryptoCell deliverables (and in fact inside regular builds of Mbed TLS as well)

Having HMAC_DRBG as the "only option" is a use-case limited to TF-M minimal configuration (hence the override of the default).

ENTROPY_CC3XX should not select PSA_WANT_ALG_CTR_DRBG. The presence of ENTROPY_CC3XX is only a statement "I want to provide entropy".  The choice of PRNG isn't made by enabling/disabling this.

Vge0rge replied:
Since we do enable the CTR_DRBG in nrfxlib by default: 
https://github.com/nrfconnect/sdk-nrfxlib/blob/c72e91d8014fe257fa4112664d213d80ae30c8c6/nrf_security/Kconfig#L189

And we also enable entropy when nrf_security is enabled: 
https://github.com/nrfconnect/sdk-nrfxlib/blob/c72e91d8014fe257fa4112664d213d80ae30c8c6/nrf_security/Kconfig#L37

Then we do we explicitly need these overrides though?  Which corner case we are trying to solve? 
I removed these from Kconfig.nrf and built some crypto samples in secure/non-secure targets and these Kconfig were always enabled.

frkv replied:
Ah. The MBEDTLS_CTR_DRBG_ENABLED is a copy of configurations from Zephyr. This should not be default y

This isn't even a valid Mbed TLS configuration (it is Zephyr only)

joerchan replied:
Moving the Kconfig overrides from `nrf/Kconfig.nrf` to `nrfxlib/nrf_security/Kconfig.psa` could be an option though.
The `!TFM_MINIMAL` seems to work better with that setup. The mbedtels_minimal.defconfig file then is able to default it to n.

SebastianBoe commented on modules/tfm/tfm/boards/common/tfm_hal_platform.c:
```diff
@@ -27,20 +31,22 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
 	int err;
 
+#if !defined(TFM_CRYPTO_RNG_MODULE_DISABLED)
 	/* Initialize the nrf_cc3xx runtime */
```
When this comment is within this ifdef it sounds like the runtime is conditionally initialized, but it is unconditional, so it should be one line higher.

SebastianBoe commented on modules/tfm/tfm/boards/common/tfm_hal_platform.c:
```diff
@@ -27,20 +31,22 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
 #if defined(TFM_PARTITION_CRYPTO)
```
Let's move crypto initialization into it's own function so that tfm_hal_platform_init reads like:

```
enum tfm_hal_status_t status;
	status = tfm_hal_platform_common_init();
	if (status != TFM_HAL_SUCCESS) {
		return status;
	}

#if defined(TFM_PARTITION_CRYPTO)
	status = crypto_init();
	if (status != TFM_HAL_SUCCESS) {
		return status;
	}
#endif

return TFM_HAL_SUCCESS;
```

When practical, a function should do things at a consistent abstraction level. (either only low-level details, or only high level details). Here we do high-level, and then low-level.

SebastianBoe commented on modules/tfm/tfm/boards/common/tfm_hal_platform.c:
```diff
@@ -27,20 +31,22 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
 #if defined(TFM_PARTITION_CRYPTO)
 	int err;
 
+#if !defined(TFM_CRYPTO_RNG_MODULE_DISABLED)
```
Let's remove this ! and flip the order to avoid the double-negative.

When I read a if(! foo),

I don't expect anything interesting to be happening when foo, because then it would have been written if(foo) else

joerchan replied:
These are TF-Ms configuration and naming. So while I agree I don't want to change this.

SebastianBoe replied:
I'm not proposing to change the name of the config.

I'm proposing to remove the !

and then change the if-else order.

joerchan replied:
Makes sense. Updated.

SebastianBoe commented on modules/tfm/tfm/boards/common/tfm_hal_platform.c:
```diff
@@ -27,20 +31,22 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
+#elif defined (PSA_WANT_ALG_HMAC_DRBG)
+	err = nrf_cc3xx_platform_init_hmac_drbg();
+#else
+	#error Please enable either PSA_WANT_ALG_CTR_DRBG or PSA_WANT_ALG_HMAC_DRBG
```
Missing " ?

SebastianBoe commented on modules/tfm/tfm/boards/common/tfm_hal_platform.c:
```diff
@@ -27,20 +31,22 @@ enum tfm_hal_status_t tfm_hal_platform_init(void)
+#endif
+#else /* !defined(TFM_CRYPTO_RNG_MODULE_DISABLED) */
+	err = nrf_cc3xx_platform_init_no_rng();
+#endif
```
To determine which nrf_cc3xx_platform_init* function to use we are using 2 PSA defines, and one TF-M define.

Is there any PSA define for a disabled RNG ?

SebastianBoe replied:
We need to clean this up with the same refactoring as mentioned here:

https://github.com/nrfconnect/sdk-nrfxlib/pull/645/files#r809971529

This should at the very least be done for the 2 PSA deifnes. It is less obvious to me if it should also apply to no_rng.

DatGizmo commented:
With the new Integration Test Specification process we don't need the github labels.
I will remove them.

Right now this PR runs all Integration tests because it changes sdk-zephyr and sdk-nrfxlib in west.yml and there is no test-spec.yml in that repo yet.
You can add changes to this PR [nrfconnect/sdk-zephyr#767](https://github.com/nrfconnect/sdk-zephyr/pull/767) and this PR [nrfconnect/sdk-nrfxlib#687](https://github.com/nrfconnect/sdk-nrfxlib/pull/687) to make the test spec here more targeted

To get the test spec feature working correctly on this PR, you need to rebase it to the latest main to get the [.github/test-spec.yml](https://github.com/nrfconnect/sdk-nrf/blob/main/.github/test-spec.yml) file

Then run CI here to update the test spec.

[Information about the test spec feature](https://projecttools.nordicsemi.no/confluence/display/NCS/Test+Spec)
[CI-Help for issues](https://ci-help.nordicsemi.no/)

joerchan commented:
@tejlmand @frkv @Vge0rge I've rebased after the upmerge and moved the PSA_WANT_ALG overrides for CTR_DRBG and HMAC_DRBG from `nrf/Kconfig.nrf` to `nrfxlib/nrf_security/Kconfig.psa`.

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-ble_mesh | X | | |
| test-fw-nrfconnect-chip | X | | |
| test-fw-nrfconnect-nfc | X | | |
| test-fw-nrfconnect-nrf-iot_libmodem | X | | |
| test-fw-nrfconnect-nrf-iot_lwm2m | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_positioning | X | | |
| test-fw-nrfconnect-nrf-iot_samples | X | | |
| test-fw-nrfconnect-nrf-iot_thingy91 | X | | |
| test-fw-nrfconnect-nrf_crypto | X | | |
| test-fw-nrfconnect-rpc | X | | |
| test-fw-nrfconnect-tfm | X | | |
| test-sdk-find-my | X | | |
| test-sdk-homekit | X | | |

**test-fw-nrfconnect-tfm:** added because there was no .github/test-spec.yml in 'trusted-firmware-m'

[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/22007/artifact/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*
